### PR TITLE
Use mocha for server tests

### PR DESCRIPTION
**Pull request by ngaruko**
# Description

[description]

medic/cht-core#[number]

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: tests/utils.js [hunk 1]
```diff
@@ -466,6 +474,22 @@ const prepServices = async (config) => {
   await runAndLogApiStartupMessage('User contact doc setup', setUserContactDoc);
 };
 
+//non browser tests
+const prepServicesNative = async () => {
```
Comment (dianabarsan) [inline tests/utils.js:478]: I think we shouldn't just add all this duplicated code, and simply refactor the code that already exists. 
Also "native" isn't an appropriate suffix here, it doesn't communicate what the difference is between this function and `prepServices`.

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Nice work!! I left a couple of recommendations inline.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -94,6 +94,83 @@ jobs:
...
+    needs: build
+    name: e2e Integration Tests for Node version ${{ matrix.node-version }}
+    runs-on: ubuntu-18.04
+
+    strategy:
+      matrix:
+        node-version: [8.x, 10.x, 12.x]
+
+    steps:
+    - name: Get Docker Hub username
```
[the remark this answers is not in the document]
Comment (newtewt) [reply .github/workflows/build.yml:107]: I looked into reusable steps and sadly it's not supported yet. We could rearchitect our builds to use composite steps but that requires them either being shell or JavaScript code. Which isn't bad.

File: tests/integration/hooks.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+const utils = require('../utils');
+const request = require('request-promise-native');
+
+exports.mochaHooks = {
+  beforeAll: async ()=> {
+    console.log('Starting services......');
+    await utils.prepServicesNative();
```
[the remark this answers is not in the document]
Comment (ngaruko) [reply tests/integration/hooks.js]: The `beforeAll` in this root hook runs only ONCE at the beginning of the test suite. In the spec, `before` is the one that runs the before all tests.

File: tests/integration/api/controllers/export-data-2.js [hunk 4]
```diff
@@ -53,13 +55,13 @@ describe('Export Data V2.0', () => {
...
-        expect(rows[0]).toBe(expected[0]);
+        expect(rows.length).to.be(expected.length);
+        expect(rows[0]).to.be(expected[0]);
         rows.splice(1).forEach(row => {
-          expect(expected).toContain(row);
+          expect(expected).to.contain(row);
         });
-      }));
-    it('POST Filters by form', () =>
+      }).then(done()));
```
[the remark this answers is not in the document]
Comment (ngaruko) [reply tests/integration/api/controllers/export-data-2.js:63]: for some reason Im getting timeout errors until I add `done` on this and some other specs

[the remark this answers is not in the document]
Comment (garethbowen) [reply tests/integration/api/controllers/export-data-2.js:63]: I think you need to make sure you return the promise. I think if you add a return on the first line of the test it'll just work, eg:

```
return utils.request({ path: '/api/v2/export/reports' }).then(result => {
```

File: .github/workflows/build.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (dianabarsan) [reply .github/workflows/build.yml:107]: Voting for Javascript to be more platform agnostic.

File: tests/utils.js [hunk 5]
```diff
@@ -261,8 +262,11 @@ const revertDb = async (except, ignoreRefresh) => {
   const needsRefresh = await revertSettings();
   await deleteAll(except);
   await revertTranslations();
+  if(ignoreRefresh){
+    await setUserContactDoc();
```
Comment (dianabarsan) [inline tests/utils.js:266]: Umm setting the userContactDoc doesn't make a lot of sense here. How come this is getting unset, or needs to be set? These docs should be always skipped on the "mass" deletion.

File: tests/utils.js [hunk 6]
```diff
@@ -459,7 +464,12 @@ const prepServices = async (config) => {
   }
 
   await listenForApi();
-  config = config || await browser.getProcessedConfig();
+  if(config === 'mocha'){
+    await listenForApi();
+    await runAndLogApiStartupMessage('User contact doc setup', setUserContactDoc);
+    return;
+  }
```
Comment (dianabarsan) [inline tests/utils.js:471]: I'm not sure I understand why this is needed. Why do we need to `listenForApi()` again? 
Also, the last line in this file is to set the user contact doc. I think it should work just fine without this condition.

File: tests/utils.js [hunk 7]
```diff
@@ -24,6 +24,7 @@ const userSettings = require('./factories/cht/users/user-settings');
 let originalSettings;
 const originalTranslations = {};
 let e2eDebug;
+const hasModal = async () => await element(by.css('#update-available')).isPresent();
```
Comment (dianabarsan) [inline tests/utils.js:27]: You don't need async + await here. returning the promise is enough: 
```suggestion
const hasModal = () => element(by.css('#update-available')).isPresent();
```

File: tests/integration/api/server.js [hunk 8]
```diff
@@ -154,38 +153,29 @@ describe('server', () => {
...
+        body: png,
+        headers: { 'Content-Type': 'image/png' },
+        method: 'PUT'
+      };
+      const { body } = await requestWrapper(options);
+      const optionsTwo = { uri: '/sample_doc2/attach?rev='+body.rev };
+      const { res, body: bodyOne } = await requestWrapper(optionsTwo);
+      expect(res.headers[ 'content-type' ]).to.equal('image/png');
+      expect(res.headers[ 'content-encoding' ]).to.be.undefined;
+      expect(bodyOne).to.equal(png);
```
Comment (dianabarsan) [inline tests/integration/api/server.js:178]:
```suggestion
      const getAttachmentOptions = { uri: '/sample_doc2/attach?rev=' + body.rev };
      const { res, body: attachmentBody } = await requestWrapper(getAttachmentOptions);
      expect(res.headers[ 'content-type' ]).to.equal('image/png');
      expect(res.headers[ 'content-encoding' ]).to.be.undefined;
      expect(attachmentBody).to.equal(png);
```

File: tests/integration/api/server.js [hunk 9]
```diff
@@ -14,16 +15,14 @@ describe('server', () => {
       };
 
       return utils.requestOnTestDb(opts, true)
-        .then(fail)
+        .then(() => assert.fail('should have thrown'))
```
Comment (dianabarsan) [inline tests/integration/api/server.js:18]: is `assert` a global here? Can we include it in the require, so we're 100% sure which `assert` we're using?

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Awesome work!! I just left a few recommendations and questions inline.

**Review by dianabarsan — APPROVED** [review]
Awesome work!

**Pull request merged**